V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-PLATFORM/MASWE-0033.md

Reviewer: DONE

---
title: Improper Use of the Clipboard
id: MASWE-0033
alias: improper-clipboard

> requirement: "The app avoids placing sensitive data on the system clipboard."

Reviewer 2: ```suggestion
requirement: "The app uses the clipboard securely and only with user consent."
```

This had come up during another review. This is my personal suggestion for consistency in titling. I've not done a full review of this weakness.

> requirement: "The app avoids placing sensitive data on the system clipboard."

Reviewer: ```suggestion
requirement: "The app validates system clipboard input and avoids copying sensitive data to it."
```

platform: [android, ios]
profiles: [L1, L2]
threat: MAS-THREAT-0033
attacks: [MAS-ATTACK-0041]
mappings:
  masvs-v2: [MASVS-PLATFORM-1, MASVS-STORAGE-2]
  cwe: [200, 668]
  android-risks:
  - secure-clipboard-handling
  maswe-beta: [MASWE-0059]
refs:
- https://developer.android.com/develop/ui/views/touch-and-input/copy-paste#PreventingSensitiveData
- https://developer.apple.com/documentation/uikit/uipasteboard
status: new
---

## Overview

This weakness occurs when an app places sensitive data on the system clipboard, or handles clipboard content insecurely, exposing that data beyond the app's control.

> The clipboard is a shared resource: other apps can read its contents, and on some platforms clipboard content synchronizes to nearby devices via a universal clipboard. Copying sensitive data such as passwords, one-time codes, card numbers, or tokens to the clipboard, failing to mark it as sensitive, or leaving it there indefinitely can leak that data to other apps and devices.

Reviewer: ```suggestion
The clipboard is a shared resource: other apps can read its contents, and on some platforms, clipboard content synchronizes to nearby devices via a universal clipboard. Copying sensitive data such as passwords, one-time codes, card numbers, or tokens to the clipboard, failing to mark it as sensitive, or leaving it there indefinitely can leak that data to other apps and devices.
```

> The clipboard is a shared resource: other apps can read its contents, and on some platforms, clipboard content synchronizes to nearby devices via a universal clipboard. Copying sensitive data such as passwords, one-time codes, card numbers, or tokens to the clipboard, failing to mark it as sensitive, or leaving it there indefinitely can leak that data to other apps and devices.

Author: ```suggestion
The clipboard is a shared resource: other apps can read its contents, and on some platforms, clipboard content synchronizes to nearby devices (e.g. the iOS [Universal Clipboard](https://support.apple.com/en-us/102430)). Copying sensitive data such as passwords, one-time codes, card numbers, or tokens to the clipboard, failing to mark it as sensitive, or leaving it there indefinitely can leak that data to other apps and devices.
```

## Modes of Introduction

> - **Sensitive Data Copyable**: Allowing sensitive values such as passwords, one-time codes, or card numbers to be copied to the clipboard.

Author: ```suggestion
- **Sensitive Data Copyable Without User Consent**: Allowing sensitive values such as passwords, one-time codes, or card numbers to be copied to the clipboard without user consent.
```

- **Universal Clipboard Not Restricted**: Not restricting sensitive clipboard items to the local device or setting an expiration on iOS, letting them sync to other devices.
- **Clipboard Not Cleared**: Leaving sensitive content on the clipboard after it has served its purpose.
- **Untrusted Clipboard Input**: Processing pasted clipboard data without validation, even though any app can have written it.

## Impact

- **Compromise of Sensitive Data**: Attackers can read personal or financial data from the clipboard, resulting in unauthorized disclosure of user data.
- **Authentication or Authorization Bypass**: Attackers can capture copied credentials or one-time codes, resulting in unauthorized access to the user's accounts.

## Mitigations

- **Avoid the Clipboard for Secrets**: Disable copying for sensitive fields and provide secure alternatives such as auto-fill (see @MASWE-0024) so users never need to copy secrets.
- **Mark Clipboard Content as Sensitive**: When sensitive content must be copied, flag it as sensitive so the platform masks previews and treats it accordingly.

> - **Restrict Clipboard Scope and Lifetime**: Keep sensitive clipboard items local to the device, set expirations where supported, and clear the clipboard once the data has been used.

Reviewer 3: ```suggestion
- **Restrict Clipboard Exposure**: Prevent copying sensitive values when it is not required for the intended functionality. When copying is a legitimate feature, follow the platform's [secure clipboard handling guidance on Android](https://developer.android.com/privacy-and-security/risks/secure-clipboard-handling) and mark the content using [`ClipDescription.EXTRA_IS_SENSITIVE`](https://developer.android.com/reference/android/content/ClipDescription#EXTRA_IS_SENSITIVE) to obscure clipboard previews. On iOS, use pasteboard options such as [`localOnly`](https://developer.apple.com/documentation/uikit/uipasteboard/optionskey/localonly) and [`expirationDate`](https://developer.apple.com/documentation/uikit/uipasteboard/optionskey/expirationdate) to limit cross-device propagation and content lifetime. These mechanisms reduce exposure but do not provide access control over clipboard contents.
```

- **Clipboard Content Not Marked Sensitive**: Not flagging copied sensitive content as sensitive where the platform supports it (e.g. `EXTRA_IS_SENSITIVE` on Android), so previews and clipboard history show it in cleartext.
- **Universal Clipboard Not Restricted**: Not restricting sensitive clipboard items to the local device or setting an expiration on iOS, letting them sync to other devices.
- **Clipboard Not Cleared**: Leaving sensitive content on the clipboard after it has served its purpose.
- **Untrusted Clipboard Input**: Processing pasted clipboard data without validation, even though any app can have written it.

## Impact

- **Compromise of Sensitive Data**: Attackers can read personal or financial data from the clipboard, resulting in unauthorized disclosure of user data.
- **Authentication or Authorization Bypass**: Attackers can capture copied credentials or one-time codes, resulting in unauthorized access to the user's accounts.

## Mitigations

- **Avoid the Clipboard for Secrets**: Disable copying for sensitive fields and provide secure alternatives such as auto-fill (see @MASWE-0024) so users never need to copy secrets.

> - **Only copy with User Consent**: If copying is required, ask the user for consent before placing sensitive data on the clipboard, or only act as a result of a user-chosen action (e.g. a "Copy" button).

Reviewer 2: ```suggestion
- **Only Copy with User Consent**: If copying is required, ask the user for consent before placing sensitive data on the clipboard, or only act as a result of a user-chosen action (e.g. a "Copy" button).
```
